src/recommendation/src/main/scala/RankingAdapter.scala

```diff
@@ -0,0 +1,382 @@
+
+  /** @group getParam */
+  def getMinRatingsPerItem: Int = $(minRatingsPerItem)
+
+  val recommender = new EstimatorParam(this, "recommender", "estimator for selection", { x: Estimator[_] => true })
+
+  /** @group getParam */
+  def getRecommender: Estimator[_ <: Model[_]] = $(recommender)
+
+  /** @group setParam */
+  def setRecommender(value: Estimator[_ <: Model[_]]): this.type = set(recommender, value)
+
+  val k: IntParam = new IntParam(this, "k", "number of items to recommend")
+
+  /** @group getParam */
+  def getK: Int = $(k)
+
+  /** @group setParam */
+  def setK(value: Int): this.type = set(k, value)
+
+}
+
+trait RankingFunctions extends RankingParams with HasRecommenderCols {
+
+  private def filterByItemCount(dataset: Dataset[_], itemCol: String, userCol: String): DataFrame =
+    dataset
+      .groupBy(userCol)
+      .agg(col(userCol), count(col(itemCol)).alias("nitems"))
+      .where(col("nitems") >= $(minRatingsPerUser))
+      .drop("nitems")
+      .cache()
+
+  private def filterByUserRatingCount(dataset: Dataset[_], itemCol: String, userCol: String): DataFrame =
+    dataset
+      .groupBy(itemCol)
+      .agg(col(itemCol), count(col(userCol)).alias("nusers"))
+      .where(col("nusers") >= $(minRatingsPerItem))
+      .join(dataset, itemCol)
+      .drop("nusers")
+      .cache()
+
+  def filterRatings(dataset: Dataset[_], itemCol: String, userCol: String): DataFrame =
+    filterByUserRatingCount(dataset, itemCol, userCol)
+      .join(filterByItemCount(dataset, itemCol, userCol), userCol)
+
+  def prepareTestData(userColumn: String, itemColumn: String,
+    validationDataset: DataFrame, recs: DataFrame,
+    k: Int): Dataset[_] = {
+    import org.apache.spark.sql.functions.{collect_list, rank => r}
+
+    val perUserRecommendedItemsDF: DataFrame = recs
+      .select(userColumn, "recommendations." + itemColumn)
+      .withColumnRenamed(itemColumn, "prediction")
+
+    val perUserActualItemsDF = if (validationDataset.columns.contains($(ratingCol))) {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(ratingCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn, $(ratingCol))
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
+    } else {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(itemCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn)
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
```

Comment:
you should use the labelCol param instead of having hard-coded value here, so user can specify the label col